docs/spec/abci/abci.md

```diff
@@ -38,20 +38,30 @@ Finally, `Query`, `CheckTx`, and `DeliverTx` include a `Codespace string`, whose
 intended use is to disambiguate `Code` values returned by different domains of the
 application. The `Codespace` is a namespace for the `Code`.
 
-## Tags
+## Events
 
 Some methods (`CheckTx, BeginBlock, DeliverTx, EndBlock`)
-include a `Tags` field in their `Response*`. Each tag is key-value pair denoting
-something about what happened during the methods execution.
+include an `Events` field in their `Response*`. Each event contains a type and a
+list of lists containing key-value pairs denoting something about what happened
+during the method's execution.
 
-Tags can be used to index transactions and blocks according to what happened
-during their execution. Note that the set of tags returned for a block from
+Events can be used to index transactions and blocks according to what happened
+during their execution. Note that the set of events returned for a block from
 `BeginBlock` and `EndBlock` are merged. In case both methods return the same
 tag, only the value defined in `EndBlock` is used.
 
-Keys and values in tags must be UTF-8 encoded strings (e.g.
-"account.owner": "Bob", "balance": "100.0",
-"time": "2018-01-02T12:30:00Z")
+Each event has a `type` which should be unique and is meant to categorize the
```

Comment:
Ok, so technically you can have multiple events with the same `type`. I guess what I was trying to state and motivate here is that it is generally unadvisable to do so. ie. it wouldn't make sense to have two or more events with type `rewards` when you can just include all the attributes under a single `rewards` event.

Reply:
> you can just include all the attributes under a single rewards event.

Not sure I follow. Isn't the whole point of this to have eg. discrete withdraw events where each has a different set of attributes, eg. one is for DelegatorA (where attributes relate to DelegatorA and the amount they withdrew) and a different event, also of the same type, but for DelegatorB, with attributes for DelegatorB ?

Reply:
@ebuchman yes -- I believe we're talking about the same thing here.

What I'm saying is the following:

It's best to have events such as this because they're all aggregated under the same `type`:

```json
"events": [
  {
    "type": "rewards",
    "attributes": [ 
        {"delegator": "delegatorA", "amount": "...", "another_attribute": "..."},
        {"delegator": "delegatorB", "amount": "...", "another_attribute": "..."},
    ]
  }
]
```


Opposed to this:


```json
"events": [
  {
    "type": "rewards",
    "attributes": [ 
        {"delegator": "delegatorA", "amount": "...", "another_attribute": "..."}
    ]
  },
  {
    "type": "rewards",
    "attributes": [ 
        {"delegator": "delegatorB"},
        {"amount": "..."},
        {"another_attribute": "..."},
    ]
  }
]
```

Does that make sense? Or am I thinking about the UX in the wrong way?

Reply:
I believe I understand what you're referring to now. I'll update the doc(s).